lib/http/connection.rb

```diff
@@ -141,6 +149,20 @@ def start_tls(req, options)
       @socket.start_tls(req.uri.host, options[:ssl_socket_class], ssl_context)
     end
 
+    # Open tunnel through proxy
+    def send_proxy_connect_request(req)
+      return unless req.uri.https? && req.using_proxy?
+
+      req.connect_using_proxy @socket
```

Comment:
Can we set the pending flags for this like below? I can't think of a case where it's strictly necessary, but when it comes to sockets and `Timeout.timeout` I'd like to be 100% sure we're protected since these issues can be subtle and nasty if they do popup.

```ruby
def send_proxy_connect_request(req)
  return unless req.uri.https? && req.using_proxy?
  
  @pending_request = true

  req.connect_using_proxy @socket

  @pending_request = false
  @pending_response = true

  read_headers!

  if @parser.status_code == 200
    @parser.reset
    @pending_response = false
    return
  end

  @failed_proxy_connect = true
end
```